File: android/app/src/main/java/com/zylod/wholesale/session/WebProvenance.kt
package com.zylod.wholesale.session

import android.util.Log
import com.zylod.wholesale.BuildConfig
import org.json.JSONObject

/**
 * F1/F5 — RUNTIME PROVENANCE.
 *
 * The owner's binding remediation order:
 *
 *  F1  "The installed APK must be able to answer: exact build → exact web
 *       source → exact web bundle commit. A wrong/stale/unknown bundle must
 *       never be silently accepted."
 *  F5  "Add the bundle commit/version marker and make it visible in
 *       diagnostics/logging so a future owner audit can immediately prove
 *       which web bundle the device is executing."
 *
 * Mechanism (smallest reliable one — the SPA is NOT bundled into the APK):
 * every web bundle build bakes its identity (commit/version/builtAt) into
 * `window.__ZylodBundleIdentity` (scripts/generate-bundle-identity.mjs, the
 * inline <head> script, and GET /api/app/version). After every document load
 * the shell reads that identity, compares it with the commit THIS native
 * build was produced from (versionName suffix, stamped by CI) and applies the
 * acceptance policy below. Everything is logged under one tag so an owner
 * audit is a single command:
 *
 *     adb logcat -s ZylodProvenance
 *
 * Acceptance policy (F1: no silent acceptance):
 *  - identity MISSING  → blocking on EVERY build type: an unknown bundle
 *    (foreign server, redirect stub, stale pre-provenance bundle) is never
 *    rendered as if it were the app.
 *  - identity MISMATCH → blocking in RELEASE; in DEBUG the exact served
 *    identity stays visible in diagnostics (the owner's order: debug builds
 *    may use development endpoints, but the selected endpoint and the exact
 *    served bundle identity must be visible) — a dev server legitimately
 *    serves an arbitrary working-tree commit, so there it is surfaced, not
 *    silently accepted and not made unusable.
 */
object WebProvenance {

    /** The ONE log tag for every provenance line (F5: permanent visibility). */
    const val LOG_TAG = "ZylodProvenance"

    /** Identity reported by the served web bundle. */
    data class BundleIdentity(
        val commit: String,
        val shortCommit: String,
        val version: String,
        val builtAt: String,
    )

    sealed class Verdict {
        /** Served bundle is the expected bundle (or expected is undeclared). */
        data class Ok(val identity: BundleIdentity) : Verdict()

        /** A known bundle, but not the one this native build was made from. */
        data class Mismatch(val identity: BundleIdentity, val expectedCommit: String) : Verdict()

        /** No identity at all — stub/foreign/stale bundle. Never accepted. */
        data object Missing : Verdict()
    }

    /** The commit this native build was produced from (CI-stamped versionName suffix). */
    fun expectedCommit(): String? {
        val name = BuildConfig.VERSION_NAME
        val suffix = name.substringAfter('-', "")
        return suffix.takeIf { it.isNotEmpty() && it != name }
    }

    fun logAppIdentity() {
        Log.i(
            LOG_TAG,
            "app=${BuildConfig.VERSION_NAME} buildType=${if (BuildConfig.DEBUG) "debug" else "release"} " +
                "expectedCommit=${expectedCommit() ?: "undeclared"} endpoints=${BuildConfig.SERVER_ENDPOINTS.joinToString(",")}",
        )
    }

    /**
     * Parses the shell's document-verify probe result (identity object or
     * JSON string form — see WebShellScripts.documentVerifyScript). Returns
     * null when the bundle exposes no usable identity.
     */
    fun parseIdentity(raw: Any?): BundleIdentity? {
        val obj: JSONObject = when (raw) {
            is JSONObject -> raw
            is String -> runCatching { JSONObject(raw) }.getOrNull() ?: return null
            else -> return null
        }
        val commit = obj.optString("commit", "").trim()
        if (commit.isEmpty() || commit == "null") return null
        return BundleIdentity(
            commit = commit,
            shortCommit = obj.optString("shortCommit", commit.take(7)).trim().ifEmpty { commit.take(7) },
            version = obj.optString("version", "").trim(),
            builtAt = obj.optString("builtAt", "").trim(),
        )
    }

    /**
     * THE acceptance policy. [debugBuild] selects the debug rule (mismatch is
     * visible, not blocking); release enforces the exact bundle.
     */
    fun evaluate(identity: BundleIdentity?, debugBuild: Boolean): Verdict {
        val expected = expectedCommit()
        val id = identity ?: return Verdict.Missing
        return if (expected != null && !id.commit.startsWith(expected) && expected != id.shortCommit) {
            Verdict.Mismatch(id, expected)
        } else {
            Verdict.Ok(id)
        }
    }

    fun blocking(verdict: Verdict, debugBuild: Boolean): Boolean = when (verdict) {
        is Verdict.Ok -> false
        is Verdict.Mismatch -> !debugBuild
        Verdict.Missing -> true
    }

    fun logServed(endpoint: String, verdict: Verdict, debugBuild: Boolean) {
        when (verdict) {
            is Verdict.Ok ->
                Log.i(LOG_TAG, "served endpoint=$endpoint bundle=${verdict.identity.shortCommit} " +
                    "version=${verdict.identity.version} builtAt=${verdict.identity.builtAt} verdict=OK")
            is Verdict.Mismatch -> {
                val level = if (debugBuild) Log.WARN else Log.ERROR
                val action = if (debugBuild) "VISIBLE-IN-DIAGNOSTICS(non-blocking)" else "BLOCKED"
                Log.println(level, LOG_TAG, "served endpoint=$endpoint bundle=${verdict.identity.shortCommit} " +
                    "expected=${verdict.expectedCommit} verdict=MISMATCH action=$action")
            }
            Verdict.Missing ->
                Log.e(LOG_TAG, "served endpoint=$endpoint bundle=NONE verdict=UNKNOWN action=BLOCKED " +
                    "(a wrong/stale/unknown bundle is never silently accepted)")
        }
    }
}

File: ios/Zylod/Web/WebProvenance.swift
import Foundation
import os

// F1/F5 — RUNTIME PROVENANCE (iOS mirror of android session/WebProvenance.kt).
//
// Owner's binding remediation order:
//  F1 "The installed IPA must be able to answer: exact build → exact web
//      source → exact web bundle commit. A wrong/stale/unknown bundle must
//      never be silently accepted."
//  F5 "Add the bundle commit/version marker and make it visible in
//      diagnostics/logging so a future owner audit can immediately prove
//      which web bundle the device is executing."
//
// Every web bundle build bakes its identity into
// `window.__ZylodBundleIdentity` (scripts/generate-bundle-identity.mjs) and
// GET /api/app/version. After every document load the shell reads that
// identity via WebShellScripts.documentVerifyScript, compares it with the
// commit THIS native build was produced from (CFBundleShortVersionString
// suffix, stamped by CI), and applies the acceptance policy below.
//
// Owner audit is one command:
//     log stream --predicate 'category == "ZylodProvenance"'
//
// Acceptance policy (F1: no silent acceptance):
//  - identity MISSING  → blocking on EVERY configuration (unknown bundle is
//    never rendered as if it were the app);
//  - identity MISMATCH → blocking in release; in DEBUG the exact served
//    identity stays visible in diagnostics (debug builds may use development
//    endpoints, but the selected endpoint and the exact served bundle
//    identity must be visible — surfaced, never silent).

enum WebProvenance {

    /// The ONE log category for every provenance line (F5).
    static let logCategory = "ZylodProvenance"
    static let logSubsystem = "com.zylod.wholesale"

    private static let logger = Logger(subsystem: logSubsystem, category: logCategory)

    /// Identity reported by the served web bundle.
    struct BundleIdentity: Equatable {
        var commit: String
        var shortCommit: String
        var version: String
        var builtAt: String
    }

    enum Verdict: Equatable {
        /// Served bundle matches this build (or the build declares no commit).
        case ok(BundleIdentity)
        /// A known bundle, but not the one this native build was made from.
        case mismatch(identity: BundleIdentity, expectedCommit: String)
        /// No identity at all — stub/foreign/stale bundle. Never accepted.
        case missing
    }

    /// The commit this native build was produced from (CI stamps
    /// CFBundleShortVersionString = "2.4.5-<sha7>" in the artifact).
    static func expectedCommit() -> String? {
        let name = (Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String) ?? ""
        let parts = name.split(separator: "-", maxSplits: 1)
        guard parts.count == 2 else { return nil }
        let suffix = String(parts[1])
        return suffix.isEmpty ? nil : suffix
    }

    static var isDebugBuild: Bool {
        #if DEBUG
        return true
        #else
        return false
        #endif
    }

    static func logAppIdentity(candidates: [String]) {
        let type = isDebugBuild ? "debug" : "release"
        logger.info("app=\(Self.versionName()) buildType=\(type, privacy: .public) expectedCommit=\(Self.expectedCommit() ?? "undeclared", privacy: .public) endpoints=\(candidates.joined(separator: ","), privacy: .public)")
    }

    static func versionName() -> String {
        (Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String) ?? "2.4.5"
    }

    /// Parses the verify probe's identity object ({commit, shortCommit,
    /// version, builtAt}) — nil when the bundle exposes no usable identity.
    static func parseIdentity(_ raw: Any?) -> BundleIdentity? {
        guard let dict = raw as? [String: Any] else { return nil }
        guard let commit = dict["commit"] as? String, !commit.isEmpty, commit != "null" else { return nil }
        let short = (dict["shortCommit"] as? String) ?? String(commit.prefix(7))
        return BundleIdentity(
            commit: commit,
            shortCommit: short.isEmpty ? String(commit.prefix(7)) : short,
            version: (dict["version"] as? String) ?? "",
            builtAt: (dict["builtAt"] as? String) ?? "",
        )
    }

    /// THE acceptance policy. In debug builds a mismatch is visible (not
    /// blocking); release enforces the exact bundle.
    static func evaluate(_ identity: BundleIdentity?) -> Verdict {
        guard let id = identity else { return .missing }
        guard let expected = expectedCommit() else { return .ok(id) }
        if id.commit.hasPrefix(expected) || expected == id.shortCommit {
            return .ok(id)
        }
        return .mismatch(identity: id, expectedCommit: expected)
    }

    static func blocking(_ verdict: Verdict) -> Bool {
        switch verdict {
        case .ok: return false
        case .mismatch: return !isDebugBuild
        case .missing: return true
        }
    }

    static func logServed(endpoint: String, verdict: Verdict) {
        switch verdict {
        case let .ok(id):
            logger.info("served endpoint=\(endpoint, privacy: .public) bundle=\(id.shortCommit, privacy: .public) version=\(id.version, privacy: .public) builtAt=\(id.builtAt, privacy: .public) verdict=OK")
        case let .mismatch(id, expected):
            if isDebugBuild {
                logger.warning("served endpoint=\(endpoint, privacy: .public) bundle=\(id.shortCommit, privacy: .public) expected=\(expected, privacy: .public) verdict=MISMATCH action=VISIBLE-IN-DIAGNOSTICS(non-blocking)")
            } else {
                logger.error("served endpoint=\(endpoint, privacy: .public) bundle=\(id.shortCommit, privacy: .public) expected=\(expected, privacy: .public) verdict=MISMATCH action=BLOCKED")
            }
        case .missing:
            logger.error("served endpoint=\(endpoint, privacy: .public) bundle=NONE verdict=UNKNOWN action=BLOCKED (a wrong/stale/unknown bundle is never silently accepted)")
        }
    }
}
